vendor/processors/base.py

```diff
@@ -58,63 +59,126 @@ def set_invoice(self, invoice):
+        """
+        self.payment = Payment(profile=self.invoice.profile,
+                               amount=self.invoice.total,
+                               provider=self.provider,
+                               invoice=self.invoice,
+                               created=timezone.now()
+                               )
+        self.payment.result['account_number'] = self.payment_info.cleaned_data.get('card_number')[-4:]
+        self.payment.payee_full_name = self.payment_info.cleaned_data.get('full_name')
+        self.payment.payee_company = self.billing_address.cleaned_data.get('company')
+
+        billing_address = self.billing_address.save(commit=False)
+        billing_address.profile = self.invoice.profile
+        billing_address.save()
+
+        self.payment.billing_address = billing_address
+        self.payment.save()
+
+    def save_payment_transaction_result(self, payment_success, transaction_id, result_info):
+        """
+        Saves the result output of any transaction. 
+        """
+        self.payment.success = payment_success
+        self.payment.transaction = transaction_id
+        self.payment.result.update(result_info)
+        self.payment.save()
 
     def update_invoice_status(self, new_status):
+        """
+        Updates the Invoice status if the transaction was submitted.
+        Otherwise it returns the invoice to the Cart. The error is saved in 
+        the payment for the transaction.
+        """
         if self.transaction_submitted:
             self.invoice.status = new_status
         else:
             self.invoice.status = Invoice.InvoiceStatus.CART
         self.invoice.save()
 
+    def is_payment_and_invoice_complete(self):
+        """
+        If payment was successful and invoice status is complete returns True. Otherwise
+        false and no receipts should be created.
+        """
+        if self.payment.success and self.invoice.status == Invoice.InvoiceStatus.COMPLETE:
+            return True
+        return False
+    
+    def get_future_date_months(self, today, add_months):
+        """
+        Returns a datetime object with the a new added months
+        """
+        newday = today.day
+        newmonth = (((today.month - 1) + add_months) % 12) + 1
+        newyear  = today.year + (((today.month - 1) + add_months) // 12)
+        if newday > mdays[newmonth]:
+            newday = mdays[newmonth]
+        if newyear % 4 == 0 and newmonth == 2:
+            newday += 1
+        return date(newyear, newmonth, newday)
+
+    def get_future_date_days(self, today, add_days):
+        """
+        Returns a datetime object with the a new added days
+        """
+        return today + timedelta(days=add_days)
+
+    def get_month_offset(self, term_type):
+        """
+        Return the number of months depending on the term type.
+        """
+        if term_type == TermType.MONTHLY_SUBSCRIPTION:
```

Comment:
No currently we have:
```
class TermType(models.IntegerChoices):
    PERPETUAL = 0, _("Perpetual")
    SUBSCRIPTION = 10, _("Subscription")
    MONTHLY_SUBSCRIPTION = 11, _("Monthly Subscription")
    QUARTERLY_SUBSCRIPTION = 12, _("Quarterly Subscription")
    SEMIANNUAL_SUBSCRIPTION = 13, _("Semi-Annual Subscription")
    ANNUAL_SUBSCRIPTION = 14, _("Annual Subscription")
    ONE_TIME_USE = 20, _("One-Time Use")
```
 I did think about that, I think it would be good to change the order around. It could either be:
```
class TermType(models.IntegerChoices):
    PERPETUAL = 0, _("Perpetual")
    SUBSCRIPTION = 10, _("Subscription")
    MONTHLY_SUBSCRIPTION = 11, _("Monthly Subscription")
    QUARTERLY_SUBSCRIPTION = 13, _("Quarterly Subscription")
    SEMIANNUAL_SUBSCRIPTION = 16, _("Semi-Annual Subscription")
    ANNUAL_SUBSCRIPTION = 12 _("Annual Subscription")
    ONE_TIME_USE = 20, _("One-Time Use")
```
Or easier :
```
class TermType(models.IntegerChoices):
    SUBSCRIPTION = 0, _("Subscription")
    MONTHLY_SUBSCRIPTION = 1, _("Monthly Subscription")
    QUARTERLY_SUBSCRIPTION = 3, _("Quarterly Subscription")
    SEMIANNUAL_SUBSCRIPTION = 6, _("Semi-Annual Subscription")
    ANNUAL_SUBSCRIPTION = 12, _("Annual Subscription")
    PERPETUAL = 20, _("Perpetual")
    ONE_TIME_USE = 30, _("One-Time Use")
```

Reply:
Yeah.  Should look to do it sooner rather than later if it makes things easier in the code.  I'd put Perpetual and One time to 100 & 110 just to give more room to the subscriptions.  So one time purchase are 100+, subscriptions are < 100